lib/media/drm_engine.js

```diff
@@ -1643,19 +1722,37 @@ shaka.media.DrmEngine = class {
   async closeOpenSessions_() {
     // Close all open sessions.
     const openSessions = Array.from(this.activeSessions_.keys());
-    this.activeSessions_.clear();
 
     // Close all sessions before we remove media keys from the video element.
     await Promise.all(openSessions.map(async (session) => {
-      shaka.log.v1('Closing session', session.sessionId);
+      const metadata = this.activeSessions_.get(session);
 
       try {
-        await this.closeSession_(session);
+        /**
+         * Special case when a persistent-license session has been initiated,
+         * without being registered in the offline sessions at start-up.
+         * We should remove the session to prevent it from being orphaned after
+         * the playback session ends
+         */
+        if (!this.offlineSessionIds_.includes(session.sessionId) &&
+          metadata.type === 'persistent-license') {
+          shaka.log.v1('Removing session', session.sessionId);
+
+          await session.remove();
+        } else {
+          shaka.log.v1('Closing session', session.sessionId, metadata);
+
+          await this.closeSession_(session);
+        }
       } catch (error) {
         // Ignore errors when closing the sessions. Closing a session that
         // generated no key requests will throw an error.
+
+        shaka.log.error('Failed to close or remove the session', error);
       }
     }));
+
+    this.activeSessions_.clear();
```

Comment:
Moving this to the end creates a possibility for a race.  If new sessions are added to activeSessions_ while the async code above is running, they won't be in openSessions, and they will be lost when we reach this point.

Please move this back to the original position.

Reply:
Done !

Reply:
Committed in https://github.com/google/shaka-player/pull/2750/commits/9d8fcdab3610fefaa8c620e3ae9fd87c4000b0c8